File: code/csharp/DesignPatterns/composite/Playlist.cs
namespace Composite;
using System.Collections.Generic;

class Playlist : IPlaylistComponent
{
    private List<IPlaylistComponent> _components = 
                                                new List<IPlaylistComponent>();

    public void AddComponent(IPlaylistComponent component)
    {
        _components.Add(component);
    }

    public void RemoveComponent(IPlaylistComponent component)
    {
        _components.Remove(component);
    }

    public void PlaySong()
    {
        System.Console.WriteLine("Playing playlist:");
        foreach (var component in _components)
        {
            component.PlaySong();
        }
    }
}

File: code/typescript/src/composite/Playlist.ts
import { PlaylistComponent } from "./PlaylistComponent";

export class Playlist implements PlaylistComponent {
    private components: PlaylistComponent[] = [];

    addComponent(component: PlaylistComponent): void {
        this.components.push(component);
    }

    removeComponent(component: PlaylistComponent): void {
        const index = this.components.indexOf(component);
        if (index !== -1) {
            this.components.splice(index, 1);
        }
    }

    playSong(): void {
        console.log("Playing playlist:");
        this.components.forEach(component => {
            component.playSong();
        });
    }
}
